app/gui/src/model/suggestion_database.rs

```diff
@@ -27,6 +29,108 @@ pub use example::Example;
 
 
 
+// ============================
+// === QualifiedNameToIdMap ===
+// ============================
+
+/// A map from [`entry::QualifiedName`]s to [`entry::Id`]s. The methods of the type provide
+/// semantics of a map, while the internal representation is based on a [`HashMapTree`].
+///
+/// The internal representation conserves memory when storing many paths sharing a common prefix of
+/// segments.
+#[derive(Clone, Debug, Default)]
+struct QualifiedNameToIdMap {
+    tree: HashMapTree<entry::QualifiedNameSegment, Option<entry::Id>>,
+}
+
+impl QualifiedNameToIdMap {
+    /// Gets the [`entry::Id`] at `path` or [`None`] if not found.
+    pub fn get<P, I>(&self, path: P) -> Option<entry::Id>
+    where
+        P: IntoIterator<Item = I>,
+        I: Into<entry::QualifiedNameSegment>, {
+        self.tree.get(path).and_then(|v| *v)
+    }
+
+    /// Sets the `id` at `path`. Emits a warning if an `id` was set at this `path` before the
+    /// operation.
+    pub fn warn_if_exists_and_set(&mut self, path: &entry::QualifiedName, id: entry::Id) {
```

Comment:
Minor thing - I'd name it `set_and_warn_if_existed` because then after the first word you understand what is the mostimportant thing this function does. In your name, you have to read all the words to get it, because the most important word is the last one.